compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -0,0 +1,332 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "AMDAIELogicalObjFifoSplittingUtils.h"
+
+#include "llvm/ADT/DenseMap.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Iterators.h"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+/// Utility to verify that the split dimensions for L2 are contiguous.
+static LogicalResult verifySplitDimensionConstraint(
+    SmallVector<unsigned> &splitDimsSetForL2) {
+  unsigned dim = 0;
+  for (unsigned splitDim : splitDimsSetForL2) {
+    if (splitDim != dim) return failure();
+    ++dim;
+  }
+  return success();
+}
+
+/*
+  For L3 -> L2 DmaCpyNd :-
+  From offset (0,0) we are extracting one 4x4 memref.
+                  _______
+                |. . . .|
+                |. . . .|
+                |. . . .|
+                |. . . .|
+                ---------
+
+  After split we will extract four 2x2 memrefs.
+  So, the corresponding offsets will be :-
+  1. Offset (0,0) - extract 2x2 memref
+        ___
+      |. .|. .
+      |. .|. .
+      -----
+        . . . .
+        . . . .
+  2. Offset (0,2) - extract 2x2 memref
+            ___
+        . .|. .|
+        . .|. .|
+          -----
+        . . . .
+        . . . .
+  3. Offset (2,0) - extract 2x2 memref
+        . . . .
+        . . . .
+        ___
+      |. .|. .
+      |. .|. .
+      -----
+  4. Offset (2,2) - extract 2x2 memref
+        . . . .
+        . . . .
+            ___
+        . .|. .|
+        . .|. .|
+          -----
+
+  The following utility helps perform the computation of offsets for L3 source.
+*/
+static FailureOr<OpFoldResult> updateL3SourceOffset(IRRewriter &rewriter,
+                                                    OpFoldResult oldL3Offset,
+                                                    int64_t offsetToAdd,
+                                                    MLIRContext *context) {
+  OpFoldResult newL3AsSourceOffset;
+  if (auto l3SourceOffsetAttr = dyn_cast<Attribute>(oldL3Offset)) {
+    int64_t l3SourceOffsetIntVal =
+        cast<IntegerAttr>(l3SourceOffsetAttr).getInt();
+    int64_t newOffset = l3SourceOffsetIntVal + offsetToAdd;
+    newL3AsSourceOffset = rewriter.getIndexAttr(newOffset);
+  } else {
+    auto l3SourceOffsetVal = cast<Value>(oldL3Offset);
+    if (auto blockArg = dyn_cast<BlockArgument>(l3SourceOffsetVal)) {
+      Operation *ownerOfBlockArg = blockArg.getOwner()->getParentOp();
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointToStart(blockArg.getOwner());
+      AffineExpr affineExpr = rewriter.getAffineDimExpr(0);
+      AffineExpr newAffineExpr = affineExpr + offsetToAdd;
+      auto newAffineMap = AffineMap::get(/*dimCount=*/1, /*symbolCount=*/0,
+                                         {newAffineExpr}, context);
```

Comment:
You could create a utility (lambda) for this as it's duplicated below in the else condition.